### PR TITLE
merge pipelines to master

### PR DESCRIPTION
**Pull request by sangyx**

### CODE REVIEW

**Comment by hameerabbasi** [comment]
You will need to fix the tests. 😦 But I'm sure you can do it! 🥳

**Comment by hameerabbasi** [comment]
If you have issues, don't hesitate to ask.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -10,8 +10,8 @@ jobs:
     displayName: Prepare conda
   - script: |
       source activate uarray
-      pip install git+https://github.com/Quansight-Labs/uarray.git
-      pip install git+https://github.com/Quansight-Labs/unumpy.git
```
Comment (hameerabbasi) [inline azure-pipelines.yml]: This should really not be commented. `udiff` should work with `master` of `uarray` and `unumpy`.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -25,31 +25,31 @@ jobs:
...
+#   - script: |
+#       source activate uarray_min
+#       python setup.py build
+#       python setup.py install
+#       pytest
+#     displayName: Run tests
+#   - task: PublishCodeCoverageResults@1
+#     inputs:
+#       codeCoverageTool: Cobertura
+#       summaryFileLocation: "$(System.DefaultWorkingDirectory)/**/coverage.xml"
```
Comment (hameerabbasi) [inline azure-pipelines.yml]: Same here.

**Review by hameerabbasi — COMMENTED** [review]
I think the tests have more-or-less the same failures, but a couple of comments.

File: azure-pipelines.yml (hunk 1 above)
Comment (sangyx) [reply azure-pipelines.yml]: OK. I will recover it in the next commit. `nbval` seems to have some problems in azure.

Comment (hameerabbasi) [reply azure-pipelines.yml]: Ouch. Let's remove the tests for now.

**Comment by sangyx** [comment]
Hi, I encountered some errors that did not appear when installing uarray and unumpy with conda. When I run code such as:
```python
import uarray as ua
import unumpy as np
import udiff
from unumpy import numpy_backend

if __name__ == "__main__":
    with ua.set_backend(numpy_backend), ua.set_backend(udiff, coerce=True):
        x = np.array([[1, 2], [3, 4]])
        y = np.exp2(x)
        y.backward()
        print(x.diff)
```

There is an error:
```
BackendNotImplementedError
No selected backends had an implementation for this function.
  File "/home/sangyx/project/gsoc/udiff/src/udiff/_vjp_diffs.py", line 168, in <lambda>
    defvjp(np.exp2, lambda ans, x: lambda g: ans * np.log(2) * g)
  File "/home/sangyx/project/gsoc/udiff/src/udiff/_vjp_core.py", line 39, in <genexpr>
    return tuple(vjp(g) for vjp in vjps)
  File "/home/sangyx/project/gsoc/udiff/src/udiff/_vjp_core.py", line 39, in ret
    return tuple(vjp(g) for vjp in vjps)
  File "/home/sangyx/project/gsoc/udiff/src/udiff/_diff_array.py", line 113, in backward
    diffs = list(self._vjp(grad_variables))
  File "/home/sangyx/project/gsoc/udiff/src/test.py", line 10, in <module>
    y.backward()
```

**Comment by hameerabbasi** [comment]
Can you try `unumpy` and `uarray` from `master` to see if you can reproduce the error?

**Comment by sangyx** [comment]
> Can you try `unumpy` and `uarray` from `master` to see if you can reproduce the error?

I install them by

````
pip install git+https://github.com/Quansight-Labs/uarray.git
pip install git+https://github.com/Quansight-Labs/unumpy.git
```


While if I use `git clone` and install them by `python setup install`, there are errors:


```
===================================== test session starts =====================================
platform linux -- Python 3.7.7, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /home/sangyx/project/gsoc/udiff, inifile: pytest.ini
plugins: cov-2.10.0, mypy-0.6.2, black-0.3.10, nbval-0.9.5
collected 0 items / 1 error                                                                   
Coverage.py warning: No data was collected. (no-data-collected)
WARNING: Failed to generate report: No data to report.

/home/sangyx/miniconda3/envs/uarray/lib/python3.7/site-packages/pytest_cov/plugin.py:271: PytestWarning: Failed to generate report: No data to report.

  self.cov_controller.finish()

=========================================== ERRORS ============================================
________________________________ ERROR collecting test session ________________________________
../../../miniconda3/envs/uarray/lib/python3.7/site-packages/_pytest/config/__init__.py:495: in _importconftest
    return self._conftestpath2mod[key]
E   KeyError: PosixPath('/home/sangyx/project/gsoc/udiff/uarray/uarray/conftest.py')

During handling of the above exception, another exception occurred:
../../../miniconda3/envs/uarray/lib/python3.7/site-packages/py/_path/local.py:704: in pyimport
    __import__(modname)
uarray/uarray/__init__.py:115: in <module>
    from ._backend import *
uarray/uarray/_backend.py:5: in <module>
    from . import _uarray
E   ImportError: cannot import name '_uarray' from 'uarray' (/home/sangyx/project/gsoc/udiff/uarray/uarray/__init__.py)
--------- generated xml file: /home/sangyx/project/gsoc/udiff/junit/test-results.xml ----------

----------- coverage: platform linux, python 3.7.7-final-0 -----------

=================================== short test summary info ===================================
ERROR  - ImportError: cannot import name '_uarray' from 'uarray' (/home/sangyx/project/gsoc/...
!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================== 1 error in 0.39s =======================================
```

**Comment by sangyx** [comment]
> Can you try `unumpy` and `uarray` from `master` to see if you can reproduce the error?

Yes. If I install them by `git clone` and then `python setup.py install`. The error is still there.

**Comment by sangyx** [comment]
The error message in pytest:
```
________________ test_unary_function[unumpy.numpy_backend-func3-<lambda>-None] ________________

backend = <module 'unumpy.numpy_backend' from '/home/sangyx/miniconda3/envs/uarray/lib/python3.7/site-packages/unumpy-0.5.0+80.g78f3323-py3.7.egg/unumpy/numpy_backend.py'>
func = <ufunc 'exp2'>, y_d = <function <lambda> at 0x7fe0bdd52a70>, domain = None

    @pytest.mark.parametrize(
        "func, y_d, domain",
        [
            (np.positive, lambda x: 1, None),
            (np.negative, lambda x: -1, None),
            (np.exp, lambda x: pow(e, x), None),
            (np.exp2, lambda x: pow(2, x) * log(2), None),
            (np.log, lambda x: 1 / x, (0, None)),
            (np.log2, lambda x: 1 / (x * log(2)), (0, None)),
            (np.log10, lambda x: 1 / (x * log(10)), (0, None)),
            (np.log1p, lambda x: 1 / (x + 1), (-1, None)),
            (np.sqrt, lambda x: 0.5 * pow(x, -0.5), (0, None)),
            (np.square, lambda x: 2 * x, None),
            (np.reciprocal, lambda x: -1 / pow(x, 2), (None, 0)),
            (np.sin, lambda x: cos(x), None),
            (np.cos, lambda x: -sin(x), None),
            (
                np.tan,
                lambda x: 1 / cos(x) ** 2,
                (-5, 5),
            ),  # Set bound to prevent numerical overflow
            (np.arcsin, lambda x: 1 / sqrt(1 - x ** 2), (-1, 1)),
            (np.arccos, lambda x: -1 / sqrt(1 - x ** 2), (-1, 1)),
            (np.arctan, lambda x: 1 / (1 + x ** 2), None),
            (np.sinh, lambda x: cosh(x), None),
            (np.cosh, lambda x: sinh(x), (1, None)),
            (np.tanh, lambda x: 1 / cosh(x) ** 2, (-1, 1)),
            (np.arcsinh, lambda x: 1 / sqrt(1 + x ** 2), None),
            (np.arccosh, lambda x: 1 / sqrt(-1 + x ** 2), (1, None)),
            (np.arctanh, lambda x: 1 / (1 - x ** 2), (-1, 1)),
            (np.absolute, lambda x: 1 if x > 0 else -1, None),
            (np.fabs, lambda x: 1 if x > 0 else -1, None),
            (np.reciprocal, lambda x: -1 / x ** 2, (1, 10)),
            (np.expm1, lambda x: exp(x), None),
            # (np.rad2deg, lambda x: 1 / pi * 180.0, None),
            # (np.deg2rad, lambda x: pi / 180.0, None),
        ],
    )
    def test_unary_function(backend, func, y_d, domain):
        if domain is None:
            x_arr = generate_test_data()
        else:
            x_arr = generate_test_data(a=domain[0], b=domain[1])
        y_d_arr = [y_d(xa) for xa in x_arr]
        try:
            with ua.set_backend(backend), ua.set_backend(udiff, coerce=True):
                x = np.asarray(x_arr)
                ret = func(x)
>               ret.backward()

src/tests/test_builtin_diffs.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/udiff/_diff_array.py:113: in backward
    diffs = list(self._vjp(grad_variables))
src/udiff/_vjp_core.py:39: in ret
    return tuple(vjp(g) for vjp in vjps)
src/udiff/_vjp_core.py:39: in <genexpr>
    return tuple(vjp(g) for vjp in vjps)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

g = array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

>   defvjp(np.exp2, lambda ans, x: lambda g: ans * np.log(2) * g)
E   uarray.BackendNotImplementedError: No selected backends had an implementation for this function.

src/udiff/_vjp_diffs.py:168: BackendNotImplementedError
```

**Comment by hameerabbasi** [comment]
This will require a refactor on the `uarray` side to solve. 😦 I'll do it once I'm back from vacation. Don't worry, it won't affect your grade.

**Pull request merged**